src/bsoncxx/types/value.hpp

```diff
@@ -154,8 +155,14 @@ namespace types {
 
         ///
         /// @return The type of the underlying BSON value stored in this object.
+        /// If the type is invalid, a disengaged optional is returned.
         ///
-        bsoncxx::type type() const;
+        stdx::optional<bsoncxx::type> type() const;
+
+        ///
+        /// @return Whether the underlying BSON value is engaged.
+        ///
+        BSONCXX_INLINE explicit operator bool() const { return type() ? true : false; }
```

Comment:
Might be better as : `return static_cast<bool>(type())` since optional already defines an explicit conversion to bool operator.